kiwi/container/lxd.py

```diff
@@ -0,0 +1,268 @@
+        root directory path name
+    """
+    def __init__(self, root_dir):
+        self.root_dir = root_dir
+
+    def create(self, filename):
+        """
+        Create LXD unified-tarball style artifact (as opposed to the
+        split-tarball style).
+
+        See https://github.com/lxc/lxd/blob/master/doc/image-handling.md#unified-tarball
+        for further detail.  This single style affords us to leave the
+        existing ContainerBuilder interface/signatures in place and also
+        helps the user in having a single artifact to manage.  It also
+        happens to be the recommended format from LXD folks too.
+
+        At the time of writing, this unified tarball style takes the
+        following shape:
+
+        rootfs/  <-- where we must put our "self.root_dir"
+        metadata.yaml
+        templates/ (optional)
+
+        Note that metadata.yaml and templates are to be found in the
+        <description_dir>/lxd_metadata/ directory.  If it doesn't exist
+        a minimal specification (dynamically created) is used.
+
+        The metadata.yaml configuration is documented at:
+        https://github.com/lxc/lxd/blob/master/doc/image-handling.md#content
+
+        Be aware that we (kiwi) assume all ownership of the two required
+        elements of metadata.yaml ("architecture: <...>" and
+        "creation_date: <...>".  Existing metadata follows these
+        dynamically constructed elements.
+
+        :param string filename: target file to create
+        """
+        # NB: we support two styles of constructing our LXD tarball:
+        # (1) the legacy mode works with all python versions but
+        #     requires an extra copy of the root_dir (yuck!)
+        # (2) the new mode works with py v3.3+ (built with lzma
+        #     support)
+        #
+        # Note also that this "different branches for different
+        # versions of Python" can give us coverage trouble given the
+        # current setup of tox/pytest arguments.  Except that our
+        # testing will be more unit-style and less end-to-end.
+        # Another solution to this (and is probably a Good Idea
+        # regardless) would be to rearchitect our tox/pytest setup to
+        # combine coverage for every environment and use that number
+        # in measuring our threshold compliance.
+        creation_method = self._get_creation_method()
+        if creation_method == 'new':
+            self._create_new(filename)
+        else:
+            self._create_legacy(filename)
+
+    def _get_creation_method(self):
+        if sys.version_info >= (3, 3) and self._has_lzma():
+            return 'new'
+        else:
+            return 'legacy'
+
+    def _has_lzma(self):  # pragma: no cover
+        try:              # ... bit tricky to cover missing imports
+            import lzma
+            lzma.is_check_supported(lzma.CHECK_NONE)  # to remedy "unused import" chirp
+            return True
+        except:
+            return False
```

Comment:
It can be shorter.

```python
try
    import lzma
except ImportError as err:
    lzma = None

...

# Remove this:
def _get_creation_method(self):
    pass

def create(self, filename):
    getattr(self, "_create_{0}".format(sys.version_info > (3, 3) and lzma is not None and 'new' or 'legacy'))(filename)
```